lib/dispatcher/dispatcher.js

```diff
@@ -25,19 +25,17 @@ class Dispatcher extends EventEmitter {
       }
 
       if (typeof interceptor !== 'function') {
-        throw new Error('invalid interceptor')
+        throw new TypeError(`invalid interceptor, expected function received ${typeof interceptor}`)
       }
 
       dispatch = interceptor(dispatch)
 
       if (dispatch == null || typeof dispatch !== 'function' || dispatch.length !== 2) {
-        throw new Error('invalid interceptor')
+        throw new TypeError('invalid interceptor')
       }
     }
 
-    this.dispatch = dispatch
-
-    return this
+    return Object.setPrototypeOf({ dispatch }, this)
```

Comment:
Also, I don't think this will properly forward close/destroy? Should we have some kind of reference count?

Reply:
>This is a bit weird? Why not just make a WrappedDispatcher class?

Hmm, that can be slightly hard, as each Dispatcher will have its own custom prototype implementation, although everyone extends from `Dispatcher`. Preserving the `prototype` is the most important thing, how do you propose to use the `WrappedDispatcher`?


>Also, I don't think this will properly forward close/destroy? Should we have some kind of reference count?

It seems like not, missed that, tests are failing most likely due to that; I'll check that out.
But what do you have in mind with reference count?
Make the `Dispatcher` aware of the number of instances wrapped?

Reply:
```
class WrappedDispatcher extends Dispatcher {
  #dispatch
  #dispatcher

  constructor(dispatch, dispatcher) {
    this.#dispatch = dispatch
  }

  dispatch(...args) {
    this.#dispatch(...args)
  }

  close (...args) {
    return this.#dispatcher.close(...args)
  }

  destroy (...args) {
    return this.#dispatcher.destroy(...args)
  }
}
```